OutOfRange error on node length 13

Hi,

Just got a runtime error (Index out of range) on the node length 13

I tried to use this on a node list of 208 separate content elements, it got stuck at trying to calculate a list of 13 nodes. I then tried adding a list of 13 nodes to the test case (copy-paste in the original tests) and got the same error.

```
~/workspace/go/src/github.com/cbergoon/merkletree [master] $ go test merkle_tree_test.go merkle_tree.go
--- FAIL: TestNewTree (0.00s)
panic: runtime error: index out of range [recovered]
	panic: runtime error: index out of range

goroutine 5 [running]:
testing.tRunner.func1(0xc42009a0f0)
	/usr/local/go/src/testing/testing.go:711 +0x2d2
panic(0x1119540, 0x11e5950)
	/usr/local/go/src/runtime/panic.go:491 +0x283
command-line-arguments.buildIntermediate(0xc420058740, 0x7, 0x8, 0x0)
	~/workspace/go/src/github.com/cbergoon/merkletree/merkle_tree.go:104 +0x50f
command-line-arguments.buildIntermediate(0xc42009c100, 0xe, 0x10, 0x20)
	~/workspace/go/src/github.com/cbergoon/merkletree/merkle_tree.go:118 +0x4e6
command-line-arguments.buildWithContent(0x11e9e20, 0xd, 0xd, 0xc4200864b0, 0xc42000a0c0, 0x4, 0x4, 0x0, 0x0)
	~/workspace/go/src/github.com/cbergoon/merkletree/merkle_tree.go:94 +0x2ff
command-line-arguments.NewTree(0x11e9e20, 0xd, 0xd, 0x11d5d20, 0x20, 0x20)
	~/workspace/go/src/github.com/cbergoon/merkletree/merkle_tree.go:63 +0x4c
command-line-arguments.TestNewTree(0xc42009a0f0)
	~/workspace/go/src/github.com/cbergoon/merkletree/merkle_tree_test.go:112 +0x86
testing.tRunner(0xc42009a0f0, 0x1146248)
	/usr/local/go/src/testing/testing.go:746 +0xd0
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:789 +0x2de
FAIL	command-line-arguments	0.010s
```

Would be a good idea to check for this case.

--- a/merkle_tree.go
+++ b/merkle_tree.go
@@ -88,7 +88,12 @@ func buildWithContent(cs []Content) (*Node, []*Node, error) {
 		})
 	}
 	if len(leafs)%2 == 1 {
-		leafs = append(leafs, leafs[len(leafs)-1])
+		duplicate := &Node{
+			Hash: leafs[len(leafs)-1].Hash,
+			C:    leafs[len(leafs)-1].C,
+			leaf: true,
+		}
+		leafs = append(leafs, duplicate)
 		leafs[len(leafs)-1].dup = true
 	}
 	root := buildIntermediate(leafs)
@@ -101,16 +106,20 @@ func buildIntermediate(nl []*Node) *Node {
 	var nodes []*Node
 	for i := 0; i < len(nl); i += 2 {
 		h := sha256.New()
-		chash := append(nl[i].Hash, nl[i+1].Hash...)
+		var left, right int = i, i+1
+		if i+1 == len(nl) {
+			right = i
+		}
+		chash := append(nl[left].Hash, nl[right].Hash...)
 		h.Write(chash)
 		n := &Node{
-			Left:  nl[i],
-			Right: nl[i+1],
+			Left:  nl[left],
+			Right: nl[right],
 			Hash:  h.Sum(nil),
 		}
 		nodes = append(nodes, n)
-		nl[i].Parent = n
-		nl[i+1].Parent = n
+		nl[left].Parent = n
+		nl[right].Parent = n
 		if len(nl) == 2 {
 			return n
 		}

--- a/merkle_tree_test.go
+++ b/merkle_tree_test.go
@@ -64,6 +64,26 @@ var table = []struct {
 	{
 		contents: []Content{
 			TestContent{
+				x: "Hello",
+			},
+			TestContent{
+				x: "Hi",
+			},
+			TestContent{
+				x: "Hey",
+			},
+			TestContent{
+				x: "Greetings",
+			},
+			TestContent{
+				x: "Hola",
+			},
+		},
+		expectedHash: []byte{46, 216, 115, 174, 13, 210, 55, 39, 119, 197, 122, 104, 93, 144, 112, 131, 202, 151, 41, 14, 80, 143, 21, 71, 140, 169, 139, 173, 50, 37, 235, 188},
+	},
+	{
+		contents: []Content{
+			TestContent{
 				x: "123",
 			},
 			TestContent{
@@ -89,6 +109,38 @@ var table = []struct {
 			},
 		},
 		expectedHash: []byte{30, 76, 61, 40, 106, 173, 169, 183, 149, 2, 157, 246, 162, 218, 4, 70, 153, 148, 62, 162, 90, 24, 173, 250, 41, 149, 173, 121, 141, 187, 146, 43},
+	},
+	{
+		contents: []Content{
+			TestContent{
+				x: "123",
+			},
+			TestContent{
+				x: "234",
+			},
+			TestContent{
+				x: "345",
+			},
+			TestContent{
+				x: "456",
+			},
+			TestContent{
+				x: "1123",
+			},
+			TestContent{
+				x: "2234",
+			},
+			TestContent{
+				x: "3345",
+			},
+			TestContent{
+				x: "4456",
+			},
+			TestContent{
+				x: "4456",
+			},
+		},
+		expectedHash: []byte{75, 116, 113, 234, 23, 149, 100, 110, 172, 129, 124, 248, 163, 158, 148, 160, 243, 161, 32, 227, 175, 254, 74, 214, 146, 64, 185, 124, 116, 137, 108, 37},
 	},
 }
 

